precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,1808 @@
+
+				var supply *big.Int
+				err = s.precompile.UnpackIntoInterface(&supply, erc20.TotalSupplyMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(supply).To(Equal(expSupply), "expected different supply")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+				Entry(" - through erc20 contract", erc20Call),
+				Entry(" - through erc20 v5 contract", erc20V5Call),
+				Entry(" - through erc20 v5 caller contract", erc20V5CallerCall),
+			)
+
+			DescribeTable("it should return zero if no tokens exist", func(callType CallType) {
+				sender := s.keyring.GetKey(0)
+				txArgs, supplyArgs := s.getTxAndCallArgs(callType, contractsData, erc20.TotalSupplyMethod)
+
+				_, ethRes, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, supplyArgs, passCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				var supply *big.Int
+				err = s.precompile.UnpackIntoInterface(&supply, erc20.TotalSupplyMethod, ethRes.Ret)
+				Expect(err).ToNot(HaveOccurred(), "failed to unpack result")
+				Expect(supply.Int64()).To(BeZero(), "expected zero supply")
+			},
+				Entry(" - direct call", directCall),
+				Entry(" - through contract", contractCall),
+				Entry(" - through erc20 contract", erc20Call),
+				Entry(" - through erc20 v5 contract", erc20V5Call),
+				Entry(" - through erc20 v5 caller contract", erc20V5CallerCall),
+			)
+		})
+
+		When("transferring tokens", func() {
+			DescribeTable("it should transfer tokens to a non-existing address", func(callType CallType, expGasUsed int64) {
+				sender := s.keyring.GetKey(0)
+				receiver := utiltx.GenerateAddress()
+				fundCoins := sdk.Coins{sdk.NewInt64Coin(s.tokenDenom, 300)}
+				transferCoins := sdk.Coins{sdk.NewInt64Coin(s.tokenDenom, 100)}
+
+				// Fund account with some tokens
+				s.fundWithTokens(callType, contractsData, sender.Addr, fundCoins)
+
+				// Transfer tokens
+				txArgs, transferArgs := s.getTxAndCallArgs(callType, contractsData, erc20.TransferMethod, receiver, transferCoins[0].Amount.BigInt())
+
+				transferCheck := passCheck.WithExpEvents(erc20.EventTypeTransfer)
+
+				res, _, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, transferArgs, transferCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				s.ExpectBalancesForContract(
+					callType, contractsData,
+					[]ExpectedBalance{
+						{address: sender.AccAddr, expCoins: fundCoins.Sub(transferCoins...)},
+						{address: receiver.Bytes(), expCoins: transferCoins},
+					},
+				)
+
+				Expect(res.GasUsed).To(Equal(expGasUsed), "expected different gas used")
+			},
+				// FIXME: The gas used on the precompile is much higher than on the EVM
+				Entry(" - direct call", directCall, int64(3_021_572)),
+				Entry(" - through erc20 contract", erc20Call, int64(54_381)),
+				Entry(" - through erc20 v5 contract", erc20V5Call, int64(52_122)),
```

Comment:
We'll have to address the gas usage still, as it is right now, the call the to EVM extension is using **way more** gas than the call to the Solidity contract.